### PR TITLE
[FLINK-2720] support flink-storm metrics

### PR DESCRIPTION
**Pull request by RalphSu**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by greghogan** [comment]
Is there a Jira ticket for this PR?

**Comment by RalphSu** [comment]
Referring to https://github.com/apache/flink/pull/1157. Not sure how to have two PR for one jira? Any suggestion

**Comment by greghogan** [comment]
Multiple PRs are added to one ticket but the commit header must reference the Jira ID. I think the comments here are sufficient so there is no need to close and open a new PR.

File: flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/CounterMetricAdapter.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import backtype.storm.metric.api.CountMetric;
+import backtype.storm.metric.api.IMetric;
+
+/**
+ * An adapter to compose the counter metric between flink counter and storm counter metric.
+ * 
+ * @since Mar 5, 2017
+ *
+ */
+public class CounterMetricAdapter extends CountMetric implements IMetric, Counter, IMetricAdapter<CountMetric> {
```
Comment (mjsax) [inline flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/CounterMetricAdapter.java]: Why does this class inherit from `Counter`? I seems to be sufficient that it as a member `Counter`.

File: flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/CounterMetricAdapter.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.metrics.Metric;
+
+import backtype.storm.metric.api.CountMetric;
+import backtype.storm.metric.api.IMetric;
+
+/**
+ * An adapter to compose the counter metric between flink counter and storm counter metric.
+ * 
+ * @since Mar 5, 2017
```
Comment (mjsax) [inline flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/CounterMetricAdapter.java]: Nit: remove this line. (also in other classes)

File: flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/CounterMetricAdapter.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ */
+public class CounterMetricAdapter extends CountMetric implements IMetric, Counter, IMetricAdapter<CountMetric> {
+
+    /** the current count */
+    // private long count;
```
Comment (mjsax) [inline flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/CounterMetricAdapter.java]: Nit: remove this.

File: flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/CounterMetricAdapter.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    private Counter delegate;
+
+    public CounterMetricAdapter(Counter counter) {
```
Comment (mjsax) [inline flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/CounterMetricAdapter.java]: nit: add `final` to all parameter (do this for all methods)

File: flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/IMetricAdapter.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import org.apache.storm.metric.api.IMetric;
+
+
+/**
+ * An adapter interface that returns storm and flink metric instance that back eath other.
+ * 
+ * @since Mar 25, 2017
+ *
+ */
+public interface IMetricAdapter<T extends IMetric> {
```
Comment (mjsax) [inline flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/IMetricAdapter.java]: I don't understand the purpose of this interface. IMHO, we don't need it.

File: flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/MetricConvert.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+	@SuppressWarnings("unchecked")
+	public static <T extends IMetric> IMetricAdapter<T> convertFlinkAdapter(String name, T originalMetric,
+			StreamingRuntimeContext context) {
+		if (originalMetric instanceof CountMetric) {
+			Counter flinkc = context.getMetricGroup().counter(name);
+			return (IMetricAdapter<T>) new CounterMetricAdapter(flinkc);
+		} else if (originalMetric instanceof MultiCountMetric) {
+			return (IMetricAdapter<T>) new MultiCountMetricAdapter(context);
+		}
+		return null;
```
Comment (mjsax) [inline flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/MetricConvert.java]: I guess it's better to throw an exception here.

File: flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/MultiCountMetricAdapter.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+import backtype.storm.metric.api.MultiCountMetric;
+
+/**
+ * @since Mar 5, 2017
+ *
+ */
+public class MultiCountMetricAdapter extends MultiCountMetric implements Metric, IMetricAdapter<MultiCountMetric> {
+
+    private Map<String, CounterMetricAdapter> _value = new HashMap<>();
```
Comment (mjsax) [inline flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/MultiCountMetricAdapter.java]: Nit: rename `_value` -> `counterPerKey`

**Comment by mjsax** [comment]
@RalphSu Do you have an JIRA id so we can assign the ticket to you?

File: flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/FlinkTopologyContext.java [hunk 8]
```diff
@@ -115,14 +126,16 @@ public ReducedMetric registerMetric(final String name, final IReducer combiner,
...
-	 * @throws UnsupportedOperationException
-	 * 		at every invocation
+	 * @return - null when storm metric not supported by flink
 	 */
 	@Override
 	public <T extends IMetric> T registerMetric(final String name, final T metric, final int timeBucketSizeInSecs) {
-		throw new UnsupportedOperationException("Metrics are not supported by Flink");
+		IMetricAdapter<T> adapter = MetricConvert.convertFlinkAdapter(name, metric, context);
+        if (adapter == null) {
+            LOG.info("Metric not supportted by Flink");
```
Comment (zentol) [inline flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/FlinkTopologyContext.java]: typo: supportted -> supported

File: flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/CounterMetricAdapter.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+    public void incr() {
+        delegate.inc();
+    }
+
+    public void incrBy(long incrementBy) {
+        delegate.inc(incrementBy);
+    }
+
+    public Object getValueAndReset() {
```
Comment (zentol) [inline flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/metric/CounterMetricAdapter.java]: Why does this return Object and not long?

**Comment by RalphSu** [comment]
update with latest master. Along fix the comments above.

**Comment by zentol** [comment]
The storm-compatibility layer will be removed in FLINK-10509. Closing this PR.

**Pull request closed without merging**